akka-docs-dev/rst/java/stream-customize.rst

```diff
@@ -9,6 +9,201 @@ is sometimes necessary to define new transformation stages either because some f
+(:class:`Outlet`). This handler will receive events related to the lifecycle of the port. In our case we need to
+override ``onPull()`` which indicates that we are free to emit a single element. There is another callback,
+``onDownstreamFinish()`` which is called if the downstream cancelled. Since the default behavior of that callback is
+to stop the stage, we don't need to override it. In the ``onPull`` callback we simply emit the next number.
+
+Instances of the above :class:`GraphStage` are subclasses of ``Graph<SourceShape<Int>,Unit>`` which means
+that they are already usable in many situations, but do not provide the DSL methods we usually have for other
+:class:`Source` s. In order to convert this :class:`Graph` to a proper :class:`Source` we need to wrap it using
+``Source.fromGraph`` (see :ref:`composition-java` for more details about graphs and DSLs). Now we can use the
+source as any other built-in one:
+
+.. includecode:: ../../../akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/GraphStageDocTest.java#simple-source-usage
+
+Port states, AbstractInHandler and AbstractOutHandler
+-----------------------------------------------------
+
+In order to interact with a port (:class:`Inlet` or :class:`Outlet`) of the stage we need to be able to receive events
+and generate new events belonging to the port. From the :class:`GraphStageLogic` the following operations are available
+on an output port:
+
+* ``push(out,elem)`` pushes an element to the output port. Only possible after the port has been pulled by downstream.
+* ``complete(out)`` closes the output port normally.
+* ``fail(out,exception)`` closes the port with a failure signal.
+
+
+The events corresponding to an *output* port can be received in an :class:`AbstractOutHandler` instance registered to the
+output port using ``setHandler(out,handler)``. This handler has two callbacks:
+
+* ``onPull()`` is called when the output port is ready to emit the next element, ``push(out, elem)`` is now allowed
+  to be called on this port.
+* ``onDownstreamFinish()`` is called once the downstream has cancelled and no longer allows messages to be pushed to it.
+  No more ``onPull()`` will arrive after this event. If not overridden this will default to stopping the stage.
+
+Also, there are two query methods available for output ports:
+
+* ``isAvailable(out)`` returns true if the port can be pushed.
+* ``isClosed`` returns true if the port is closed. At this point the port can not be pushed and will not be pulled anymore.
+
+The relationship of the above operations, events and queries are summarized in the state machine below. Green shows
+the initial state while orange indicates the end state. If an operation is not listed for a state, then it is invalid
+to call it while the port is in that state. If an event is not listed for a state, then that event cannot happen
+in that state.
+
+|
```

Comment:
inserts a blank line. I use it to pad the images otherwise they clamp up with the text above and below. This of course should be handled by the CSS, but I did not dare to touch that.